text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,334 @@
+["peer dependencies"](https://nodejs.org/es/blog/npm/peer-dependencies) (see
+[Appendix](#peer-dependencies-in-npm) for details). 
+
+This means that consumers of a 3rd-party construct library are expected to
+**directly** install all transitive dependencies. 
+
+For example, if someone publishes a 3rd-party construct library `foo` that uses the `@aws-cdk/aws-ecs`, the consumers of this library will not only have to take a dependency on `foo`, but also on `@aws-cdk/aws-ecs` **and all it's transitive dependencies** (i.e. 23 additional dependencies).
+
+This is an unacceptable user experience with our current tools.
+
+## Poor Ergonomics
+
+In addition to the peer npm dependency issue described above, due to the fact
+that CDK is "hyper modular", in all languages, we have poor ergonomics when it
+comes to declaring dependencies. Since users are required to explicitly install
+a module for each service they use, even simple projects end up with dozens of
+direct CDK dependencies.
+
+# Basic Example
+
+The AWS CDK will be shipped as a single module that includes the core types and the entire AWS Construct Library.
+
+This means that a 3rd-party library will declare it's dependency on the CDK via a single module `aws-cdk-lib` (name to be decided):
+
+```json
+{
+  "name": "your-awesome-lib",
+  "peerDependencies": {
+    "aws-cdk-lib": "^2.12.0",
+    "constructs": "^1.0.0"
+  }
+}
+```
+
+> The `constructs` module includes the core programming model of the CDK, which we plan to extract into a separate library.
+
+And that's it. Apps that consume this library will depend on it and will also
+depend on the CDK:
+
+```json
+{
+  "name": "my-awesome-app",
+  "dependencies": {
+    "aws-cdk-lib": "2.89.0",
+    "constructs": "1.77.0",
+    "your-awesome-lib": "^1.0.0"
+  }
+}
+```
+
+In JavaScript code, import statements that use the CDK will now look like this:
+
+```ts
+import { aws_s3, aws_dynamodb } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+
+// ...
+export class MyConstruct extends Construct {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    new aws_s3.Bucket(this, ...);
+    new aws_dynamodb.Table(this, ...);
+  }
+}
+```
+
+> TODO: other languages
+
+# Design Summary
+
+The general approach proposed by this RFC is to ship the core types and the
+entire AWS Construct Library as a single module for all languages.
+
+This will dramatically simplify how users declare their compatibility with the
+AWS CDK, and I would argue it is also more aligned with our user's mental model (see [Rational](#rationale-and-alternatives) beflow).
+
+# Detailed Design
+
+## Prototype
+
+See [monocdk-experiment](https://www.npmjs.com/package/monocdk-experiment).
+
+## TODO
+
+- [ ] We lose per-module analytics which means we will to move to report
+  analytics at the construct level.
+- [ ] We jsii to support the concept of "submodules". I believe this is easier
```

Comment:
What was the problem with supporting namespaces in jsii again?

IIRC, the only REAL issue was that in TypeScript, a class can be both a class and a namespace and that's all fine, but that's hard to translate to other languages (i.e., trouble translating `CfnBlaBucket.EncryptionProperty`).

Why are namespaces hard in general? All languages we're targeting (with the possible exception of Go?) support them.

Reply:
There isn't an intrinsic issue with namespaces, just that it was hard to find a good way to represent their full semantics in all languages in a consistent way. For example, in Python we ran into issues related to cyclic dependencies between types defined inside the namespace and outside the namespace.

I would argue that it will be "safer" to add more explicit support for "submodules" in jsii, and then we can decide how to represent them in each language.

@RomainMuller can surely expand on that

Reply:
The main issue we have faced with namespaces was that certain compositions that are perfectly workable in JavaScript cannot be represented in Python. This is in parts due to how interfaces have no physical existence in JavaScript, whereas we materialize them in Python... So if the cyclic dependency is solely caused by interfaces, the TypeScript compiler can work it out... But Python suddenly cannot get out of the cycle.

Reply:
See [prototype](https://github.com/aws/jsii/pull/1297)